### ISSUE TITLE
Cancel function feature undocumented

### ISSUE THREAD

**Issue by epoberezkin** [jut-io/bluebird-retry#9]
It also can create some unexpected consequences if somebody passes the function that returns the promise but also calls the callback if it's supplied.

Not sure how the cancel feature can be used at all, I guess to prevent retrying on certain types of errors. The better approach may have been to pass an optional `cancelWhen` (or `retryWhen`) function via options that would be checking the promise rejection and returning `true` (or `false`) if the retrying should be cancelled and the opposite value otherwise.

In any case, documenting this feature could be helpful... I can do it if you like, just let me know which approach you prefer.

**Comment by demmer** [comment]
Yeah I agree -- the current cancel API is a bit obscure and I hadn't thought of that problem when we implemented it the first time.

Thinking about it a bit more, the better approach to me would be to have the retry function expose an Error subclass to signal that iteration should stop. That way a client could do something like:

```
var retry = require('bluebird-retry');

retry(function() {
    if (/* something fatal */) {
        throw new retry.CancelError('stop retrying');
    }  else if (/* not fatal, but not success yet */) {
        throw new Error('keep trying');
    } else {
        return 'success';
    }
});
```

**Comment by epoberezkin** [comment]
Yes, I like it

**Comment by dcollinsf5** [comment]
+1 for this; I implemented this exactly myself in my current project, for a case where I know that one type of error will eventually work itself out and succeed, but other error types are critical failures that would simply be wasting time retrying.  It's especially useful since, in my case, the error type that will eventually work itself out may take 3 minutes to get there, whereas other failures should reject immediately (or close to it)

**Comment by demmer** [comment]
How does #11 look?

**Comment by dcollinsf5** [comment]
I like it!  Just something you might like/not be aware of - check out bluebird's .catch() syntax; it's extremely nice for this type of thing.

Bluebird's standard way of dealing with cancellable promises is to add a catch like so:
.catch(bluebird.CancellationError, function(err) { / handle cancel here/ })

This works because you can use the prototype of an Error object as the first argument to bluebird's .catch(), to catch cleanly, without having to branch inside your catch() function to identify the type of error provided.
@ line 60:
.catch(function(err) {
        if (err instanceof CancelError) {
        if (err.err instanceof Error) {
        return Promise.reject(err.err);
        } else {
        return Promise.reject(err);
        }
        }

Could be 2 catches
.catch(CancelError, function(err) {
//logic specific to cancelError
})
.catch(function(err) {
// Logic for any other kinds of errors
)

Has the added benefit that errors thrown from your CancelError catch will automatically be caught in the second catch.

It makes catching a bunch of errors very clean:

.catch(errortype1, errortype1handler)
.catch(errortype2, errortype2handler)
...etc

**Comment by demmer** [comment]
Resolved with #11.

**Comment by epoberezkin** [comment]
:+1:
